### PR TITLE
[SPARK-50529][SQL] Change char/varchar behavior under the `spark.sql.preserveCharVarcharTypeInfo` config

### PR DESCRIPTION
**Pull request by jovanm-db**
### What changes were proposed in this pull request?
This PR changes char/varchar behaviour under the `PRESERVE_CHAR_VARCHAR_TYPE_INFO` configuration flag, (exposed as `spark.sql.preserveCharVarcharTypeInfo`). 

### Why are the changes needed?
This PR enables the improvement of char/varchar types in a backwards compatible way. 

### Does this PR introduce _any_ user-facing change?
No. 

### How was this patch tested?
Added tests in:
- `RowEncoderSuite`
- `LiteralExpressionSuite`
- `CharVarcharTestSuite`

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by mihailom-db** [comment]
LGTM, please update title to reflect the config name, also give a SQL name in the description, so that spark users can know which config you are adding.

**Review by dongjoon-hyun — COMMENTED** [review]
Is this a Draft status because there is no implementation, @jovanm-db ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -4818,6 +4818,14 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHAR_VARCHAR_EXTENDS_STRING = buildConf("spark.sql.charVarcharExtendsString")
+    .doc("When true, Spark treats CHAR/VARCHAR type differently than STRING type, which is the " +
+      "opposite behavior then that of Spark 3.0 and earlier. This means length check for " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4823]: `then`?

Maybe, 
```
- which is the opposite behavior then that of Spark 3.0 and earlier.
+ which is the opposite of its behavior in Spark 3.0 and earlier versions.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -4818,6 +4818,14 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHAR_VARCHAR_EXTENDS_STRING = buildConf("spark.sql.charVarcharExtendsString")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4821]: how is this different from `LEGACY_CHAR_VARCHAR_AS_STRING`?

Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4821]: Yeah, I agree this naming is not perfect. Maybe we could make clear what the config main idea is `CHAR_VARCHAR_AS_PROPER_TYPES` as we will replace char/varchar with CharType and VarcharType in the underlying code instead of using StringType.

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -4818,6 +4818,14 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val PRESERVE_CHAR_VARCHAR_TYPE_INFO = buildConf("spark.sql.preserveCharVarcharTypeInfo")
+    .doc("When true, Spark treats CHAR/VARCHAR type differently than STRING type, which is the " +
+      "opposite of its behavior in Spark 3.0 and earlier versions. This means length check for " +
+      "CHAR/VARCHAR type and padding for CHAR type when writing data to the table.")
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4824]: Please update this doc as well.

**Comment by cloud-fan** [comment]
I think it's better to add the changes that use this conf in this PR. Small PRs are good but PR should be self-contained.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 4]
```diff
@@ -86,6 +86,31 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
     }
   }
 
+  test("preserve char/varchar type info") {
+    Seq(CharType(5), VarcharType(5)).foreach { typ =>
+      withSQLConf(SQLConf.LEGACY_CHAR_VARCHAR_AS_STRING.key -> "true") {
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:91]: Could you also add the default behavior without setting any config. This IMO is a good thing to do, as that would make the test fail in future if some config default value is changed.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 5]
```diff
@@ -80,10 +80,12 @@ object RowEncoder extends DataTypeErrorsBase {
...
-      case CharType(_) | VarcharType(_) =>
-        throw new AnalysisException(
-          errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
-          messageParameters = Map("dataType" -> toSQLType(dataType)))
+      case CharType(_) | VarcharType(_) if !SqlApiConf.get.preserveCharVarcharTypeInfo =>
+          throw new AnalysisException(
+            errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
+            messageParameters = Map("dataType" -> toSQLType(dataType)))
+      case CharType(length) => CharEncoder(length)
+      case VarcharType(length) => VarcharEncoder(length)
```
Comment (mihailom-db) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:88]:
```suggestion
      case CharType(length) if SqlApiConf.get.preserveCharVarcharTypeInfo => CharEncoder(length)
      case VarcharType(length) if SqlApiConf.get.preserveCharVarcharTypeInfo => VarcharEncoder(length)
      case CharType(_) | VarcharType(_) if !SqlApiConf.get.preserveCharVarcharTypeInfo =>
          throw new AnalysisException(
            errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
            messageParameters = Map("dataType" -> toSQLType(dataType)))
```
Why I suggest this is that this is more error proof and more readable. What you say like this is that default behaviour is throw an error, which is what the behaviour was before. With your implementation I feel like this config was added to make something fail, but the default behaviour was to pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 6]
```diff
@@ -318,6 +318,11 @@ object Cast extends QueryErrorsBase {
     case _ if from == to => true
     case (NullType, _) => true
     case (_: NumericType, _: NumericType) => true
+    case (_: AtomicType, CharType(_) | VarcharType(_))
+      if !SQLConf.get.preserveCharVarcharTypeInfo => false
+    case (CharType(a), CharType(b)) => a <= b
+    case (CharType(a), VarcharType(b)) => a <= b
+    case (VarcharType(a), VarcharType(b)) => a <= b
     case (_: AtomicType, CharType(_) | VarcharType(_)) => false
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:326]: Ditto here, the code looks weird as well.
```suggestion
    case (CharType(a), CharType(b)) if SQLConf.get.preserveCharVarcharTypeInfo => a <= b
    case (CharType(a), VarcharType(b)) if SQLConf.get.preserveCharVarcharTypeInfo => a <= b
    case (VarcharType(a), VarcharType(b)) if SQLConf.get.preserveCharVarcharTypeInfo => a <= b
    case (_: AtomicType, CharType(_) | VarcharType(_)) => false
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 7]
```diff
@@ -87,7 +87,9 @@ object ExpressionEncoder {
       }
       constructProjection(row).get(0, anyObjectType).asInstanceOf[T]
     } catch {
-      case e: SparkRuntimeException if e.getCondition == "NOT_NULL_ASSERT_VIOLATION" =>
+      case e: SparkRuntimeException if
+        e.getCondition == "NOT_NULL_ASSERT_VIOLATION" ||
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:91]: nit: If I am not mistaken, from style perspective we usually leave conditions up to 100 characters and then break.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala [hunk 8]
```diff
@@ -80,6 +80,32 @@ object DeserializerBuildHelper {
...
+      length: Int): Expression = {
+    val expr = StaticInvoke(
+      classOf[CharVarcharCodegenUtils],
+      StringType,
+      "varcharTypeWriteSideCheck",
+      path :: Literal(length) :: Nil,
+      returnNullable = returnNullable)
+    createDeserializerForString(expr, returnNullable)
+  }
+
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/DeserializerBuildHelper.scala:108]: Please add tests that check that deserialization actually performs the length check.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 9]
```diff
@@ -318,6 +318,11 @@ object Cast extends QueryErrorsBase {
     case _ if from == to => true
     case (NullType, _) => true
     case (_: NumericType, _: NumericType) => true
+    case (_: AtomicType, CharType(_) | VarcharType(_))
+      if !SQLConf.get.preserveCharVarcharTypeInfo => false
+    case (CharType(a), CharType(b)) => a <= b
+    case (CharType(a), VarcharType(b)) => a <= b
+    case (VarcharType(a), VarcharType(b)) => a <= b
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:325]: Is this change necessary in this PR? I would say opening a separate PR to fix casting is better than mixing stuff.

**Review by mihailomilosevic2001 — APPROVED** [review]
I have some comments regarding the new changes. Also 2 more things. First, update the PR title and description, the main idea of the PR now is not adding the config, but the change under it. Do we need some guards on the configs we have. What I mean by this is that charVarcharAsString, charAsVarchar and preserveCharVarcharTypeInfo seem to be mutually exclusive, as in only one can be true at the time. @cloud-fan what do you think about this?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 10]
```diff
@@ -80,11 +80,11 @@ object RowEncoder extends DataTypeErrorsBase {
...
-      case CharType(_) | VarcharType(_) =>
-        throw new AnalysisException(
-          errorClass = "UNSUPPORTED_DATA_TYPE_FOR_ENCODER",
-          messageParameters = Map("dataType" -> toSQLType(dataType)))
-      case _: StringType => StringEncoder
+      case CharType(length) if SqlApiConf.get.preserveCharVarcharTypeInfo =>
+        CharEncoder(length)
+      case VarcharType(length) if SqlApiConf.get.preserveCharVarcharTypeInfo =>
+        VarcharEncoder(length)
+      case s: StringType if s.constraint == NoConstraint => StringEncoder
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:87]: if `preserveCharVarcharTypeInfo` is false, we should always convert char/varchar to string type before creating the encoder?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala [hunk 11]
```diff
@@ -63,6 +63,36 @@ object SerializerBuildHelper {
     Invoke(inputObject, "doubleValue", DoubleType)
   }
 
+  def createSerializerForChar(inputObject: Expression, length: Int): Expression = {
+    val expr = StaticInvoke(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala:67]: is it `createSerializerForString`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 12]
```diff
@@ -176,7 +176,8 @@ object TableOutputResolver extends SQLConfHelper with Logging {
     if (canWriteValue) {
       val nullCheckedValue = checkNullability(value, attr, conf, colPath)
       val casted = cast(nullCheckedValue, attrTypeWithoutCharVarchar, conf, colPath.quoted)
-      val exprWithStrLenCheck = if (conf.charVarcharAsString || !attrTypeHasCharVarchar) {
+      val exprWithStrLenCheck = if ((conf.charVarcharAsString &&
+        !conf.preserveCharVarcharTypeInfo) || !attrTypeHasCharVarchar) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:180]: why? no matter we preserver char/varchar type info or not, we should always apply length check and padding.

**Comment by cloud-fan** [comment]
In this PR we check the `preserveCharVarcharTypeInfo` flag in many places. Do we have a narrow wrist? IIUC we replace the char/varchar type with string type for all leaf scan nodes, are there any other entry points that can produce char/varchar type and need to check the flag?

**Comment by mihailom-db** [comment]
@cloud-fan Only one more that I know of is Cast expression.

**Comment by cloud-fan** [comment]
I think we only need to check the flag at these entry points. For other places, just do whatever is needed to handle char/varchar, as if they see char/varchar, it means the flag is turned on.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 10 above)
Comment (jovanm-db) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:87]: If `preserveCharVarcharTypeInfo` is false the behavior is the same as before `CharType` and `VarcharType` inherited from `StringType`, which is to throw an `AnalyssiException`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala (hunk 11 above)
Comment (jovanm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala]: Yes. I removed the duplicated code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala (hunk 12 above)
Comment (jovanm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala]: The added condition was indeed unnecessary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala [hunk 13]
```diff
@@ -373,7 +373,7 @@ object ResolveDefaultColumns extends QueryErrorsBase
           throw QueryCompilationErrors.defaultValuesDataTypeError(
             statementType, colName, defaultSQL, dataType, other.dataType))
     }
-    if (!conf.charVarcharAsString && CharVarcharUtils.hasCharVarchar(dataType)) {
+    if (!conf.charVarcharAsString || CharVarcharUtils.hasCharVarchar(dataType)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:376]: why is this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 14]
```diff
@@ -196,9 +198,13 @@ object Literal {
     case TimestampNTZType => create(0L, TimestampNTZType)
     case it: DayTimeIntervalType => create(0L, it)
     case it: YearMonthIntervalType => create(0, it)
-    case CharType(_) | VarcharType(_) =>
-      throw QueryExecutionErrors.noDefaultForDataTypeError(dataType)
-    case st: StringType => Literal(UTF8String.fromString(""), st)
+    case CharType(length) if SQLConf.get.preserveCharVarcharTypeInfo =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:201]: `Literal.default` is a very internal API, do we really need to check the flag here? If we see char/varchar (we should not if the flag is false), we should just handle it correctly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 15]
```diff
@@ -140,6 +141,7 @@ case class AnalyzeColumnCommand(
     case DoubleType | FloatType => true
     case BooleanType => true
     case _: DatetimeType => true
+    case CharType(_) | VarcharType(_) if !SQLConf.get.preserveCharVarcharTypeInfo => false
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala:144]: I don't think we are ready to support char/varchar column statistics, the story is not completed yet. E.g. how do we consider padding when calculating min/max? And how do we push down predicates since Spark does read-side padding by default? The predicate is like `rpad(col, 5) > 'a'`.

I'd prefer to keep char/varchar unsupported for now.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 16]
```diff
@@ -86,6 +95,27 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
     }
   }
 
+  test("preserve char/varchar type info") {
+    Seq(CharType(5), VarcharType(5)).foreach { typ =>
+      val char_varchar_as_string = SQLConf.LEGACY_CHAR_VARCHAR_AS_STRING.key
+      val preserve_char_varchar = SQLConf.PRESERVE_CHAR_VARCHAR_TYPE_INFO.key
+      withAllConfigCombinations(Seq(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:102]: nit:
```
for {
  charVarcharAsString <- Seq(true, false)
  preserveCharVarchar <- Seq(true, false)
} {
  withSQLConf ...
}
```

**Review by cloud-fan — APPROVED** [review]
LGTM if CI passes

**Comment by cloud-fan** [comment]
thanks, merging to master!
